node-cpp.md

```diff
@@ -1,66 +1,115 @@
+* [Why create a node addon?](#why-create-a-node-addon)
+* [How is it different than a C++ project?](#how-is-it-different-than-a-c-project)
+* [Native Abstractions for Node.js (NAN)](#native-abstractions-for-nodejs-nan)
+* [Examples](#examples)
+* [Developing addons](#developing-addons)
+* [Including other C++ headers into your project](#including-other-c-headers-into-your-project)
+* [Versioning](#versioning)
+* [Additional Resources](#additional-resources)
+
+The following document outlines Mapbox's general approach to writing C++ modules for Node.js (often referred to as _addons_), and they _why_. Check out [node-cpp-skel](https://github.com/mapbox/node-cpp-skel), a skeleton library for creating a Node.js addon, to learn more about _how_ to create an addon.
 
 Node is integral to the Mapbox APIs. Sometimes at scale, though, Node becomes a bottleneck for performance. Node is single-threaded, which blocks execution. C++ on the other hand allows you to execute operations without clogging up the event loop (learn more about the node event loop [here](https://nodejs.org/en/docs/guides/event-loop-timers-and-nexttick/)). Passing heavy operations into C++ and subsequently into C++ workers can greatly improve the overall runtime of the code.
 
-### Native Abstractions for Node.js (NAN)
+### Why create a node addon?
 
-To swing between Node and C++, the Node community maintains a project called [_NAN_](https://github.com/nodejs/nan) (Native Abstractions for Node.js) that simplifies running different versions of Node and, subsequently, v8. NAN is a header-only C++ library that provides a set of Macros for developing Node.js addons. Check out the [usage](https://github.com/nodejs/nan#usage) guidelines.
+1. To port a C++ project to Node to expose a new interface for the tool (like Mapnik & Node Mapnik)
+1. Improve performance at scale where Node becomes the bottleneck (i.e. concurrency)
 
-More examples of how to port C++ libraries to node can be found at [nodejs.org/api/addons](http://nodejs.org/api/addons.html). See https://nodesource.com/blog/c-add-ons-for-nodejs-v4/ for a detailed summary of the origins of Nan. And see http://blog.reverberate.org/2016/10/17/native-extensions-memory-management-part2-javascript-v8.html for a good introduction to v8 that is not specific to node.js.
+**Concurrency**
+
+Concurrency is the process of executing different pieces of the same process to allow for parallel execution of these pieces [[wikipedia](https://en.wikipedia.org/wiki/Concurrency_(computer_science))]. Node.js addons allow us to take advantage of concurrent operations within our node applications by reaching into more than just the v8 thread, but this can result in a few surprises. In the example below, we’ve passed data from our node application, into the v8 thread, and subsequently into our worker threadpool. Here there are multiple workers executing code at the same time. In our worker, we have the following line:
```

Comment:
Link to [v8 glossary term](https://github.com/mapbox/cpp/blob/master/glossary.md#v8)